### PR TITLE
Fix glReadPixels

### PR DESCRIPTION
**Pull request by OhadRau**
Addresses the issue I brought up in https://github.com/revery-ui/revery/issues/185. Again, this is one of those places where I feel like I sort of had to employ a hack, but it works on my test inputs (easiest way to verify that this is working is to change the glClearColor to red or blue).

Basic description of what I changed:
* I was previously setting the image descriptor byte based on what some example code was doing. Not sure why, but this swapped the y-axis so now I just set the image descriptor byte to 0 (which should work according to the TGA spec).
* I was previously copying in the raw pixel array as returned by glReadPixels. This would break if you used GL_UNSIGNED_BYTE on a little-endian system, so instead I'm swapping the red/blue channels in the glReadPixels stubs (both native & js).

### CODE REVIEW

File: src/gl_stubs.js [hunk 1]
```diff
@@ -287,5 +287,21 @@ function caml_glReadPixels_bytecode(x, y, width, height, vFormat, vType, data) {
...
 
   joo_global_object.gl.readPixels(x, y, width, height, format, type, data);
+
+  // If we're on a little-endian system, the R/B channels are swapped
+  // So let's determine endianness...
+  var marker = new ArrayBuffer(4);
+  var u32view = new Uint32Array(marker);
+  u32view[0] = 0x12345678;
+  var u8view = new Uint8Array(marker);
+  if (u8view[0] == 0x78 && type == joo_global_object.gl.UNSIGNED_BYTE) {
```
Comment (bryphe) [inline src/gl_stubs.js]: Wow, thanks for implementing the logic to determine endianness! Surprising that the JS engine doesn't handle this for us. Thanks for figuring this out 👍

File: src/gl_wrapper.cpp [hunk 2]
```diff
@@ -619,6 +619,19 @@ extern "C" {
...
 
+      // If we're on a little-endian system, the R/B channels are swapped
+      // So let's determine endianness...
+      unsigned int marker = 0x12345678;
+      if (*(char *) &marker == 0x78 && type == GL_UNSIGNED_BYTE) {
+        // We are little-endian. Onto the swap...
+        int numChannels = format == GL_RGBA ? 4 : 3;
+        for (int i = 0; i < width * height * numChannels; i += numChannels) {
+          uint8_t tmp = *((uint8_t *) data + i);
+          *((uint8_t *) data + i) = *((uint8_t *) data + i + 2);
```
Comment (bryphe) [inline src/gl_wrapper.cpp]: Fun 😄

**Comment by bryphe** [comment]
> Not sure why, but this swapped the y-axis so now I just set the image descriptor byte to 0 (which should work according to the TGA spec).

How'd you figure this out? Did you have to dive into the TGA spec? Nice find!

**Comment by OhadRau** [comment]
Lol it's actually a weird edge case I bumped into when implementing the original PR. TypedArrays respect system endianess whereas DataViews don't, which was kind of confusing to figure out. I tried to keep the C and JS as close as possible here to make sure both platforms are correct though.

**Comment by OhadRau** [comment]
Yeah, I used this http://www.paulbourke.net/dataformats/tga/. Only the second type (unmapped RGB) matters for us luckily.

Anyways I just tested this in Revery and it worked as expected. I'm wondering if there's any way to compile this in big endian on ARM so that I can test that, but it seems like such an edge case for a GUI lib that it might not be worth bothering with.

**Comment by bryphe** [comment]
> Anyways I just tested this in Revery and it worked as expected. I'm wondering if there's any way to compile this in big endian on ARM so that I can test that, but it seems like such an edge case for a GUI lib that it might not be worth bothering with.

Great! I'm on-board with this change as-is. We haven't tested ARM support at all across the board, so I suspect this wouldn't be the only issue.

**Comment by bryphe** [comment]
I'll publish a new NPM release once this is in so we can pick it up easily in `revery` 👍

**Comment by OhadRau** [comment]
In that case I guess we can merge this in now
